src/utils/usePromise.js

```diff
@@ -0,0 +1,110 @@
+/* eslint react/prop-types: 0 */
+// @flow
+import * as React from 'react';
+
+function resolvePromise(promise) {
+  if (typeof promise === 'function') {
+    return promise();
+  }
+
+  return promise;
+}
+
+const states = {
+  pending: 'pending',
+  rejected: 'rejected',
+  resolved: 'resolved',
+  initial: 'initial',
```

Comment:
Maybe we should rename these to match useAsync: loading, failure, success, initial.